src/main/java/io/fabric8/elasticsearch/plugin/acl/ACLDocumentManager.java

```diff
@@ -0,0 +1,287 @@
+import io.fabric8.elasticsearch.plugin.PluginClient;
+import io.fabric8.elasticsearch.plugin.PluginSettings;
+import io.fabric8.elasticsearch.plugin.acl.SearchGuardRoles.Roles;
+import io.fabric8.elasticsearch.plugin.acl.SearchGuardRolesMapping.RolesMapping;
+
+/**
+ * Manages process of loading and updating the ACL Documents
+ * for a user request 
+ *
+ */
+public class ACLDocumentManager implements ConfigurationSettings {
+    
+    private static final Logger LOGGER = Loggers.getLogger(ACLDocumentManager.class);
+    private final ReentrantLock lock = new ReentrantLock();
+    private final String searchGuardIndex;
+    private final PluginClient client;
+    private final SearchGuardSyncStrategyFactory documentFactory;
+
+    public ACLDocumentManager(final PluginClient client, final PluginSettings settings, final SearchGuardSyncStrategyFactory documentFactory) {
+        this.searchGuardIndex = settings.getSearchGuardIndex();
+        this.client = client;
+        this.documentFactory = documentFactory;
+    }
+    
+    @SuppressWarnings("rawtypes")
+    interface ACLDocumentOperation{
+        
+        void syncDocuments(Collection<SearchGuardACLDocument> docs);
+        
+        BulkRequest buildRequest(Client client, BulkRequestBuilder builder, Collection<SearchGuardACLDocument> docs) throws IOException;
+    }
+
+    @SuppressWarnings("rawtypes")
+    class SyncFromContextOperation implements ACLDocumentOperation {
+
+        private OpenshiftRequestContext context;
+
+        public SyncFromContextOperation(OpenshiftRequestContext context) {
+            this.context = context;
+        }
+        
+        @Override
+        public void syncDocuments(Collection<SearchGuardACLDocument> docs) {
+            LOGGER.debug("Syncing from context to ACL...");
+            for (SearchGuardACLDocument doc : docs) {
+                if(ConfigurationSettings.SEARCHGUARD_MAPPING_TYPE.equals(doc.getType())){
+                    RolesMappingSyncStrategy rolesMappingSync = documentFactory.createRolesMappingSyncStrategy((SearchGuardRolesMapping) doc);
+                    rolesMappingSync.syncFrom(context);
+                } else if(ConfigurationSettings.SEARCHGUARD_ROLE_TYPE.equals(doc.getType())) {
+                    RolesSyncStrategy rolesSync = documentFactory.createRolesSyncStrategy((SearchGuardRoles) doc);
+                    rolesSync.syncFrom(context);
+                }
+            }
+        }
+
+        @Override
+        public BulkRequest buildRequest(Client client, BulkRequestBuilder builder, Collection<SearchGuardACLDocument> docs) throws IOException{
+
+            for (SearchGuardACLDocument doc : docs) {
+                UpdateRequestBuilder update = client
+                        .prepareUpdate(searchGuardIndex, doc.getType(), SEARCHGUARD_CONFIG_ID)
+                        .setDoc(doc.toXContentBuilder());
+                if(doc.getVersion() != null) {
```

Comment:
under what circumstances would there not be a document version here?  IIUC, the document version is the key to ensuring the atomicity of updates?

Reply:
The only case I can think of is when its a 'create' operation which would mean initialization.  It also could be an artifact of testing but I am unable to concretely say.